### PR TITLE
fix(adapters): preserve late tool aliases and reject invalid indexes

### PR DESCRIPTION
**Pull request by lidge-jun**
## Summary

A streamed Chat tool call may arrive ID-only, then ID+index, then index-only. Preserve the first observed valid index as an alias so arguments remain attached to one call, while its original key continues to own translator-budget reservations.

Carries #3673 by @devswha with the original authored commit. Adds early rejection of present non-null indexes that are non-numeric, negative, fractional or unsafe before any identity fallback, plus malformed-index and raw large-integer collision regressions proving no reassignment, no successful terminal event and released reservations. Existing parallel, collision and UTF-8 budget cases remain intact.

**D stack:** #3681 documentation and #3684 TOML foundation are merged; this next layer targets dev. Subsequent dependent review layers will name this branch while it remains open. Review this layer's diff only. Close source #3673 only after dev ancestry proves integration.

## Verification

- Independent current-tree plan audit passed; original source and contributor identity verified.
- `git diff origin/dev...HEAD --check` passed.
- Independent implementation review: PASS, no actionable blockers; original authored patch preserved by stable patch ID.
- Exact head `01f91991140bce1248842a2d9ee62b6dcd1898c2`: isolated macmini-cf, project Bun 1.4.0 / Node22.22.0 passed typecheck and complete suite (19,551 pass,15 skip,0 fail), plus docs build (425 pages).
- GitHub checks are pending; no claim they passed. No local application suite/typecheck was run.

## Checklist

- [x] Scope stays focused and avoids unrelated cleanup.
- [x] Docs or release notes were updated when needed.
- [x] Security-sensitive changes were reviewed for secrets, auth, and unsafe defaults.

Co-authored-by: Hako <25837994+devswha@users.noreply.github.com>

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved streamed tool-call tracking for parallel requests, including late-arriving identity information and index-versus-ID resolution.
  - Invalid tool-call indexes—such as negative, fractional, unsafe, or unsupported values—now stop streaming with an error instead of being misinterpreted.
  - Preserved support for missing or null indexes and correct resource cleanup when streaming fails.

- **Documentation**
  - Added guidance on streamed tool-call identity handling, valid index values, and error behavior.

- **Tests**
  - Expanded coverage for identity collisions, boundary values, malformed indexes, budget handling, and UTF-8 limits.

### CODE REVIEW

File: src/adapters/openai-chat.ts [hunk 1]
```diff
@@ -1848,17 +1849,26 @@ export function createOpenAIChatAdapter(provider: OcxProviderConfig): ProviderAd
               const rawId = rawToolCall.id;
               const idDelta = typeof rawId === "string" ? rawId : "";
               const rawIndex = rawToolCall.index;
+              // Invalid numeric indexes must not fall through to ID or last-call matching.
+              // Reject before an alias can bind or any pending call can consume the fragment.
+              if (typeof rawIndex === "number"
+                  && (!Number.isInteger(rawIndex) || rawIndex < 0)) {
```
[the remark this answers is not in the document]
Comment (lidge-jun) [reply src/adapters/openai-chat.ts]: Fixed in d6bfb044a: Number.isSafeInteger rejects unsafe numeric indexes before matching. Added a raw SSE/JSON regression with distinct 9007199254740992 and 9007199254740993 literals, immediate failure and reservation release, plus a positive MAX_SAFE_INTEGER late-alias case. Independent repair review passed. Exact-head remote Bun1.4.0/Node22.22.0 typecheck and complete suite passed (19369 pass,14 skip,0 fail), with docs build425 pages.

**Comment by lidge-jun** [comment]
## 리뷰 · 우선순위 59 / 80

이 PR은 openai-chat 스트림에서 tool call이 “ID만 → ID+index → index만” 순으로 올 때, 처음 본 유효 index를 별칭으로 남겨 인자 조각이 한 호출에 붙게 합니다. 동시에 음수·소수·unsafe integer index는 매칭 전에 끊어서, 잘못된 호출에 인자가 붙거나 JS 숫자 정밀도 붕괴로 두 index가 하나로 합쳐지는 길을 막습니다. 원본 `#3673`(devswha/Hako) carry에 safe-integer 가드와 회귀 테스트를 덧붙인 D-lane 다음 층입니다. 상위 `#3681` 문서·`#3684` TOML은 이미 `dev`에 있습니다.

왜 지금 `dev`에 필요한가. 지금 `dev`의 `PendingToolCall`은 `key`만 있고, 첫 청크가 ID-only면 `key=id:…`로 열립니다. 이어지는 index-only 조각은 `key=i:N`으로 새 줄을 만들거나, index가 비정상이면 “없다”고 보고 마지막 pending에 붙일 수 있습니다. 그러면 인자가 갈라지거나 다른 호출에 섞입니다. 이 변경은 `indexKey?`를 두고, 직접 key → index 별칭 → id 순으로 찾고, 예산(`budget.openCall`) 소유 key는 처음 것으로 고정합니다. `Number.isSafeInteger`로 범위 밖 리터럴도 거부하고, `MAX_SAFE_INTEGER` 경계는 허용합니다.

Grok 스택 `#3698`–`#3703`과 무관합니다. C-lane OAuth(`#3691`/`#3688`)와도 파일이 안 겹칩니다. `#3700` 카탈로그 ordering과도 독립입니다. merge-tree 기준 코드 충돌 없고, GitHub상 `MERGEABLE`(체크 대기라 BLOCKED)입니다. types/config 분할과 무관합니다. 어댑터 안전성 레이어라 CI 초록이면 단독으로 넣어도 됩니다.

`src/adapters/openai-chat.ts` `PendingToolCall.indexKey` - ID로 시작한 호출이 나중에 index를 배우면 별칭만 붙입니다. 첫 유효 index만 고정하고, 같은 ID가 다른 index로 다시 와도 별칭을 바꾸지 않습니다.

`src/adapters/openai-chat.ts` invalid numeric index 가드 - `typeof rawIndex === "number"`이면서 safe-integer가 아니거나 `< 0`이면 `terminateWithError`로 끊습니다. ID/last-call fallback으로 떨어지지 않습니다. 예약은 `closeToolCalls`로 전부 풀립니다.

경로 조회 순서 `key` → `indexKey` → `id` - index가 있는 조각이 충돌 ID보다 이깁니다. 테스트 `an observed index wins over a conflicting ID`가 고정합니다.

경로 `tests/adapters/openai/openai-chat-parallel-stream.test.ts` - late alias, 병렬·미매칭 index-only, 음수/소수, unsafe raw JSON 리터럴 붕괴, MAX_SAFE_INTEGER, duplicate ID, UTF-8 byte limit까지 있습니다. `createTestTranslatorBudget`로 예약 해제도 확인합니다.

경로 문서 `adapters.md` / `structure/04_transports-and-sidecars.md` - 계약을 코드와 맞춰 적었습니다. plan `020`/`021`은 D-lane 기록입니다.

경로 원본 `#3673` - 아직 OPEN입니다. `dev` ancestry 확인 후 rebase 말고 landed 닫기입니다.

메인테이너의 판단이 필요한 지점
- hosted CI(지금 큐/일부 성공)가 전부 초록일 때만 바로 머지할지
- 문자열 index(`"0"`)는 기존처럼 non-numeric 관용으로 둘지, 나중에 엄격 모드를 둘지(이번 범위 밖)
- `#3673`을 머지 직후 바로 닫을지

너의 추천
D-lane 어댑터 안전 수정이고 `#3700`·Grok·OAuth와 독립입니다. **CI 초록이면 바로 머지**를 추천합니다. 머지 직후 `#3673`에 `Landed via #3702 at <commit>` + `landed-via-maintainer`로 닫으세요.

이 댓글은 grok-bot이 작성했습니다

File: src/adapters/openai-chat.ts [hunk 2]
```diff
@@ -1848,17 +1849,27 @@ export function createOpenAIChatAdapter(provider: OcxProviderConfig): ProviderAd
               const rawId = rawToolCall.id;
               const idDelta = typeof rawId === "string" ? rawId : "";
               const rawIndex = rawToolCall.index;
+              // Invalid numeric indexes must not fall through to ID or last-call matching.
+              // Unsafe integers can collapse distinct wire indexes onto the same JS number.
+              // Reject before an alias can bind or any pending call can consume the fragment.
+              if (typeof rawIndex === "number"
+                  && (!Number.isSafeInteger(rawIndex) || rawIndex < 0)) {
```
[the remark this answers is not in the document]
Comment (lidge-jun) [reply src/adapters/openai-chat.ts]: Fixed in baf8303bb / validated head01f919911: every present non-null index must be a non-negative safe integer. String/empty-string/bool/object/array cases now reject before matching, while missing/null continuation remains supported. All11 labeled invalid cases and positive placeholders passed exact-head remote full validation (19551 pass,15 skip,0 fail), with typecheck/docs425 and independent review.

File: docs-site/src/content/docs/reference/adapters.md [hunk 3]
```diff
@@ -49,6 +49,13 @@ provider — xAI, Kimi, DeepSeek, GLM, Groq, OpenRouter, Ollama (local), and mor
   tiers, accepts reasoning deltas from either `delta.reasoning_content` or `delta.reasoning`, requests
   streamed usage with `stream_options.include_usage`, and reads usage from non-stream response envelopes.
 
+Streaming tool calls retain their identity when a provider first sends an ID,
+then associates that ID with an index, and later sends index-only argument
+fragments. Those fragments assemble into one call with the original name and
+complete arguments; parallel calls retain separate identities.
+Numeric streamed tool-call indexes must be non-negative safe integers; malformed numeric
+indexes terminate the stream with an upstream error before identity matching.
```
[the remark this answers is not in the document]
Comment (lidge-jun) [reply docs-site/src/content/docs/reference/adapters.md]: Updated the public contract in baf8303bb to explicitly name non-numeric values and negative, fractional or unsafe numbers, and to preserve only missing/null absent-index placeholders. Documentation build and independent review passed at01f919911.
